### PR TITLE
Adds manageiq_user as an action_plugin

### PR DESCRIPTION
**Pull request by syncrou**
##### SUMMARY
Manageiq modules need the functionality to receive `url`, `token` and future attributes as `extra_vars` which are then passed through to the module.  This change modifies `manageiq_user` into  an action_plugin which allows the above requirement to work as expected.

##### ISSUE TYPE
 - New Plugin Pull Request

##### COMPONENT NAME
Action Plugin: manageiq_user

##### ANSIBLE VERSION
```
ansible 2.4.0 (convert_manageiq_user_to_plugin b06f8c5c59) last updated 2017/09/07 16:07:36 (GMT -400)
  config file = /Users/dbomhof/syncrou/playbooks/ansible.cfg
  configured module search path = [u'/Users/dbomhof/syncrou/ansible/lib/ansible/modules']
  ansible python module location = /Users/dbomhof/syncrou/ansible/lib/ansible
  executable location = /Users/dbomhof/syncrou/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
[ManageIQ](http://manageiq.org) allows for the creation of Ansible based services.  Essentially a clickable button that allows for a playbook to execute.  These Ansible services pass manageiq specific `api_url`, `api_token` information to the playbook via `extra_vars`.  

This change allows those specific connection based attributes to be passed through as a `module_var`  directly to the `manageiq_user` module.  This allows playbooks run from inside ManageIQ to gain authentication behavior while also simplifying the arguments required to make them run.

##### EXAMPLE COMMAND
```
ansible-playbook example_playbook.yml --extra-vars "{'manageiq':{'api_url':'http://localhost:3000', 'api_token':'sometoken'}}" -vvv
```

### CODE REVIEW

**Comment by syncrou** [comment]
@yaacov - Please review ( and add other names as you see fit )

cc - @gmcculloug, @gtanzillo, @mkanoor

**Comment by yaacov** [comment]
@joelddiaz please review

**Comment by mattclay** [comment]
bot_status

**Comment by yaacov** [comment]
@cben please review

**Comment by joelddiaz** [comment]
works as advertised. lgtm

**Review by yaacov — APPROVED** [review]
LGTM

**Comment by yaacov** [comment]
shipit

**Comment by syncrou** [comment]
bot_status

File: lib/ansible/module_utils/manageiq.py
Comment (cben) [inline lib/ansible/module_utils/manageiq.py:32]: Can we please use the same names?  Either change the extra vars manageiq is passing in, or rename the args the manageiq modules expect.

Similarly, could the magic var name be `manageiq_connection`?

File: lib/ansible/module_utils/manageiq.py
Comment (cben) [inline lib/ansible/module_utils/manageiq.py:92]: Why verify_ssl handled separately outside MANAGEIQ_MODULE_VARS?

File: lib/ansible/module_utils/manageiq.py
Comment (cben) [inline lib/ansible/module_utils/manageiq.py:73]: returns module_vars but also mutates it in-place.  please do one or the other (I prefer deepcopy and return new, if possible)

File: lib/ansible/module_utils/manageiq.py
Comment (cben) [inline lib/ansible/module_utils/manageiq.py:96]: Does this work — is it 'token' or 'api_token' at this point?

Why is this needed?
I guess it's so a playbook specifying user&pwd wins over ansible var specifying token?
IINM, doing this after the merge it can have other effects eg. ansible var specifying user&pwd wins over playbook specifying a token...

* Idea: Do we have to merge at all?  What if we simplify the behavior to "whole `manageiq_connection` param, if unset, defaults to `manageiq_connection` ansible var"?  If someone really needs to specify some params and inherit others from a var, they can do it themselves with {{...}} templating, with full control.

File: lib/ansible/modules/remote_management/manageiq/manageiq_user.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_user.py:99]: Sorry, "a native Service" says nothing to me.
Consider giving an example explicitly setting an ansible var, and explaining in a comment that happens automatically when running playbooks inside ManageIQ.

I'd like the actual mechanics documented.  Ideally for each affected param, near the existing explanation (module_docs_fragments/manageiq.py) of defaulting to MIQ_USERNAME env var etc.  Something like "if a `manageiq` ansible variable is in scope, is a hash, and contains `username` key, ...".  Including priority between env var and ansible var...

The reason I'm pressing for all this is transparency and debuggability.
It worries me this is all behind-the-scenes opaque magic to the playbook writer, and at some point will bite someone and they'll not have any idea what's going on and how to debug their playbook.
If we hold this to the high standard "If the implementation is hard to explain, it's a bad idea.", it will force us to make the behavior more predictable...

**Review by cben — CHANGES_REQUESTED** [review]
I have no idea if this is idiomatic thing to do in Ansible.
We have existing mechanism of defaulting to env vars, but if this is considered good practice, :+1: in principle.

(I'd love someone from Ansible to weigh in, and override me!)

But I want to hold it to a high standard of "public API".
Forget temporarily use case of magically working when ManageIQ executes ansible.
Consider use case of playbook writer using same mechanism — set an ansible var (by any of the myriad ways to set ansible vars) and have manageiq_* modules magically see it.
If we can document that well, manageiq can just happen to use this by passing --extra-vars...

[Sorry for the verbose review]

**Comment by cben** [comment]
Update following offline discussion with @syncrou:
Env vars only work on `hosts: localhost` when you execute ansible directly.  It'll be rare to want `hosts: another-machine` with these modules; but another use case is telling AWX (Ansible Tower) to run the module, where again you don't control env vars with which module is executed.
=> Direction of adding action plugins LGTM.

My comments about aligning the peeked vars / module args names still stand.

**Comment by bcoca** [comment]
so argspec has a way for modules to handle env vars ...we don't currently have anything for action plugins, so it is fine they do it themselves for now ... probably going to fix that soon.

**Pull request closed without merging**
